### ISSUE TITLE
failed to dial gRPC: unable to upgrade to h2c, received 502

### ISSUE THREAD

**Issue by kousu** [moby/moby#37959]
**Description**

**Steps to reproduce the issue:**
1. Install Docker for Windows or Docker for Mac, stable.
2. Run `docker build .` on a docker project

**Describe the results you received:**

```
$ docker build . -t user-manager
ERRO[0000] failed to dial gRPC: unable to upgrade to h2c, received 502 
context canceled
```

**Describe the results you expected:**

Docker to work at all.

**Additional information you deem important:**

Disabling "experimental features" and restarting the docker daemon fixes this. It's as simple as that.

This issue has been open for over a year at https://github.com/docker-library/docker/issues/71, but while plenty of users have confirmed it across a range of versions and platforms, nothing has been done. The maintainers there punted the problem, said to take it over here, but no one did. When it's everyone's responsibility, it's no one's responsibility, I guess?

Why are _experimental_ features getting shipped in the _stable_ docker channel at all? Why are they on by default? Why has this issue been open for over a year?

**Output of `docker version`:**

```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:21:31 2018
 OS/Arch:           darwin/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:29:02 2018
  OS/Arch:          linux/amd64
  Experimental:     false

```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 11
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.93-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: YYVQ:JL5S:DTBJ:GPUK:OJJ7:DXN3:VC6C:AXSV:WKKT:UH26:B62F:A3PZ
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 25
 Goroutines: 52
 System Time: 2018-10-03T07:47:43.940101873Z
 EventsListeners: 2
HTTP Proxy: gateway.docker.internal:3128
HTTPS Proxy: gateway.docker.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

VirtualBox 5.2.18r124319
OS X Mojave (10.14)

**BUT REMEMBER: this has been seen on Windows and several variants of OS X**

**Comment by kousu** [comment]
That's what https://github.com/docker-library/docker/issues/71 makes it sound like.

**Comment by AkihiroSuda** [comment]
The error code of 
https://github.com/docker-library/docker/issues/71 is 501 while this issue is 502

**Comment by kousu** [comment]
That's true, but it also gave the same text with a 404 over there, and plus the same fix (of _disabling_ experimental features) worked for me, so I assumed all the issues were related.

I am trying to make a minimal test case for you now.

**Comment by kousu** [comment]
I've turned back on experimental features on my docker and wiped all my images and containers one by one and now `docker build .` is behaving itself and I cannot reproduce the bug.

Nevertheless, this is a real issue that was biting a lot of people over in https://github.com/docker-library/docker/issues/71; that project said they thought it was probably a bug upstream (i.e. here, unless I'm mistaken?) so I am going to leave this open here to collect other reports.

I don't know for sure if this is only with the Windows/Mac installs. I followed https://docs.docker.com/v17.12/docker-for-mac/install/#install-and-run-docker-for-mac, downloading https://download.docker.com/mac/stable/Docker.dmg. Issue docker-library/docker#71 sounds like it's mostly happening to Win/Mac people, so maybe there's something weird about the GUI client that linux users aren't running into.

**Comment by thaJeztah** [comment]
@djs55 any ideas?

**Comment by swuecho** [comment]
have simliar problem. 

the problem goes away after I pull the base docker image and re-run the build command.

**Comment by iguanaus** [comment]
If it is any help, this issue is caused on Mac if you have not let the Docker Desktop completely load. See below: 

```
(fpg_venv) Johns-MacBook-Pro-3:fpg_gitlab johnpeurifoy$ docker build -t iguanaus/ebs_deploy_repository:tag .
ERRO[0000] failed to dial gRPC: unable to upgrade to h2c, received 502 
context canceled
(fpg_venv) Johns-MacBook-Pro-3:fpg_gitlab johnpeurifoy$ docker build -t iguanaus/ebs_deploy_repository:tag .
Sending build context to Docker daemon  857.1kB
....
```

**Comment by Progdrasil** [comment]
I'm having the same problem, here is the diagnostic ID

**Comment by onesimusaz** [comment]
One thing I keep needing to remember is that Windows shells run at different permissions levels. I ran into this problem when trying to run the docker sample application. 

The fix for me was simply to run my powershell terminal as Administrator. (Similar to running sudo in *nix)

**Comment by lhocke** [comment]
I was running into this but like @onesimusaz mentioned I wasn't running powershell as Administrator. Problem resolved as soon as I did so

**Comment by daveisfera** [comment]
This [happened to me on Docker for Mac](https://github.com/docker/for-mac/issues/4793) and I couldn't access anything with Docker and had to restart it to get it working again: C18223FD-44DC-4ACD-B197-B660FC33917F/20200722004247

**Comment by tomzx** [comment]
I helped diagnose a similar issue with a colleague of mine (she was on OS X) and the fix was to increase the amount of allocated RAM (in the Preferences - Resources tab). From my minimal understanding of docker, the daemon which is responsible for building the image probably ends up being OOM and dies, which terminates the connection between the server (daemon) and the client (CLI) with a status 502.

From a user perspective the experience isn't great. The error is cryptic and doesn't tell us much about what could be done. While the assessment done here might be wrong, it would be great if the error was "caught" and some kind of suggestions were given to the user, in this case "increase the amount of allocated RAM".

**Comment by hzhijun** [comment]
> Happening on linux, too. To reproduce,
> 
> git clone https://github.com/blockstack/stacks-blockchain-api.git npm install docker-compose -f docker-compose.dev.stacks-blockchain.yml build --no-cache
> 
> output:
> 
> ```
> [+] Building 0.0s (0/0)                                                                                                  
> failed to get status: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing unable to upgrade to h2c, received 404"
> ```

I also encountered the same error, did you solve your problem?

**Comment by bartoszgolebiowski** [comment]
This issue is caused by a mismatch between the client and engine versions.
I had a 1.40 client version and a 1.41 engine.

To sum up, it tried to resolve the h2c dependency (the first dependency in the graph) and had some issues connecting to the DOCKER_ENGINE via POST request; the request consists of the docker client API version (1.40), for example. 

Try to align the Docker client and Docker engine version to resolve the issue.
